renku_notebooks/api/classes/server.py

```diff
@@ -0,0 +1,399 @@
+import os
+import requests
+from kubernetes import client
+from kubernetes.client.rest import ApiException
+from hashlib import md5
+import base64
+import json
+from urllib.parse import urlparse
+from uuid import uuid4
+from flask import make_response, jsonify
+import logging
+from urllib.parse import urljoin
+
+
+from ...util.check_image import parse_image_name, get_docker_token, image_exists
+from ...util.kubernetes_ import (
+    get_k8s_client,
+    secret_exists,
+    filter_pods_by_annotations,
+)
+
+
+class UserServer:
+    """Represents a jupuyterhub session."""
+
+    def __init__(
+        self,
+        user,
+        namespace,
+        project,
+        branch,
+        commit_sha,
+        notebook,
+        image,
+        server_options,
+    ):
+        self._renku_annotation_prefix = "renku.io/"
+        self._get_environment_vars()
+        self._user = user
+        self._k8s_client, self._k8s_namespace = get_k8s_client()
+        self.safe_username = self._user.safe_username
+        self.namespace = namespace
+        self.project = project
+        self.branch = branch
+        self.commit_sha = commit_sha
+        self.notebook = notebook
+        self.image = image
+        self.server_options = server_options
+        self.server_name = self.make_server_name(
+            self.namespace, self.project, self.branch, self.commit_sha
+        )
+
+    def _get_environment_vars(self):
+        if os.environ.get("JUPYTERHUB_API_TOKEN", None) is None:
+            raise ValueError(
+                "The jupyterhub API token is missing, it must be provided in "
+                "an environment variable called JUPYTERHUB_API_TOKEN"
+            )
+        if os.environ.get("GITLAB_URL", None) is None:
+            raise ValueError(
+                "The gitlab URL is missing, it must be provided in "
+                "an environment variable called GITLAB_URL"
+            )
+        if os.environ.get("IMAGE_REGISTRY", None) is None:
+            raise ValueError(
+                "The url to the docker image registry is missing, it must be provided in "
+                "an environment variable called IMAGE_REGISTRY"
+            )
+        self._default_image = os.environ.get(
+            "NOTEBOOKS_DEFAULT_IMAGE", "renku/singleuser:latest"
+        )
+        self._image_registry = os.environ.get("IMAGE_REGISTRY")
+        self._jupyterhub_authenticator = os.environ.get(
+            "JUPYTERHUB_AUTHENTICATOR", "gitlab"
+        )
+        self._git_url = os.environ.get("GITLAB_URL")
+        self._jupyterhub_path_prefix = os.environ.get(
+            "JUPYTERHUB_BASE_URL", "/jupyterhub"
+        )
+        self._jupyterhub_origin = os.environ.get("JUPYTERHUB_ORIGIN", "")
+
+    @staticmethod
+    def make_server_name(namespace, project, branch, commit_sha):
```

Comment:
@ableuler I removed `renku_notebooks/util/jupyterhub_.py` that was not used anywhere at all. I just completely missed it.

As for making this a property I would like to keep it as a static method because it allows me to use it in the tests here:
https://github.com/SwissDataScienceCenter/renku-notebooks/pull/541/files#diff-205841df8f264252b295f3101d045d55f31a0044526da199256b89e12d9518dfR29

And even though the tests that use `make_servername` so far are not that fancy - we can write more in the future and using this would be useful. Also it truly does not depend on anything else I just thought it made sense to be part of the server class. Lastly because we use this as the name of the JH server then JH enforces this result to be unique. So by keeping it a static method we can easily see what we and JH considers to be a unique server.